hilti/toolchain/src/compiler/optimizer.cc

```diff
@@ -1618,6 +1621,57 @@ struct DeadStoreVisitor : OptimizerVisitor, visitor::PreOrder<bool, DeadStoreVis
     }
 };
 
+struct DeadCodeVisitor : OptimizerVisitor, visitor::PreOrder<bool, DeadCodeVisitor> {
+    bool prune_uses(Node& node) override {
+        bool any_modification = false;
+
+        while ( true ) {
+            bool modified = false;
+
+            for ( auto i : this->walk(&node) ) {
+                if ( auto x = dispatch(i) )
+                    modified = modified || *x;
+            }
+
+            if ( ! modified )
+                break;
+
+            any_modification = true;
+        }
+
+        return any_modification;
+    }
+
+    result_t operator()(const Module& m, position_t p) {
+        _current_module = &p.node.as<Module>();
+        return false;
+    }
+
+    result_t operator()(const statement::Expression& x, position_t p) {
+        if ( auto ctor = x.expression().tryAs<expression::Ctor>();
+             ctor && (ctor->isConstant() || ! hasChildWithSideEffects(x)) ) {
+            HILTI_DEBUG(logging::debug::Optimizer,
+                        util::fmt("removing side-effect free dead code (%s)", x.meta().location()));
+            removeNode(p);
+            return true;
+        }
+
+        return false;
+    }
+
+    bool hasChildWithSideEffects(const Node& node) const {
+        auto v = visitor::PreOrder<>();
+
+        // NOLINTNEXTLINE(readability-use-anyofallof)
+        for ( const auto& child : v.walk(node) ) {
+            if ( child.node.isA<operator_::struct_::MemberCall>() || child.node.isA<operator_::function::Call>() )
```

Comment:
You seem to have reviewed an older version, I believe this approach should be fine. There is the assumption that operators which do not compile to HILTI, but are emitted directly do not have side effects beyond the type they operate on (e.g., mutate global, observable state). This seems to hold for the operators I looked into.

As an example consider `+=` for `bytes` (maps onto `Bytes::append` when emitting C++). Since this operator only changes internal state of a `bytes` instance it is enough to check how this instance is used, e.g., if it is local the function can be pure, but not so for a global.